app/src/main/java/org/mozilla/fenix/library/historymetadata/controller/HistoryMetadataGroupController.kt

```diff
@@ -119,16 +120,28 @@ class DefaultHistoryMetadataGroupController(
         scope.launch {
             items.forEach {
                 store.dispatch(HistoryMetadataGroupFragmentAction.Delete(it))
-                activity.components.core.historyStorage.deleteHistoryMetadata(it.historyMetadataKey)
+                activity.components.core.historyStorage.deleteVisitsFor(it.url)
                 metrics.track(Event.HistorySearchTermGroupRemoveTab)
             }
+            // The method is called for both single and multiple items.
+            // In case all items have been deleted, we have to disband the search group.
+            if (store.state.items.isEmpty()) {
```

Comment:
Since `store.dispatch` runs in its own scope (using a single-threaded dispatcher, by default - and we don't override this default), this `if` statement will race against the dispatch above.

To avoid this, I think it's worth refactoring these methods so that we have an explicit codepath for when we need to dispatch a `disband` action, and not try to infer that based on the state.

Otherwise, here you're basically doing what a middleware may do, except in a buggy way.

Reply:
We could simple compare the size of existing items and items to remove, and if they match, consider it to be a full removal. Right?

We have an explicit `handleDeleteAll` method, which is called when the remove all button is clicked. 
But `handleDelete` could also lead to a case when the last item is deleted: all items were selected and deleted, or user was deleting them one by one. It looks like the controller is the exact place (not interactor, not fragment) where we have to decide if the UI is trying to remove the last item, and it is going to be based on the state. Or how we can do that otherwise?

We could move it to middleware, but HistoryMetadataGroupFragmentStore is a separate middleware and knows nothing about the rest. The controller, on the other side, has access to `components.core.store`, so it looks pretty natural to allow it to emit events. No strong opinion though.